src/main/java/cceclipseplugin/core/PluginManager.java

```diff
@@ -135,65 +165,304 @@ public DocumentManager getDocumentManager() {
-											.getRequest(null, null));
-									}
-								} catch (ConnectException e) {
-									e.printStackTrace();
-								}
-
-								// For creating new file
-								// IRequestData data = new
-								// FileCreateRequest(StringConstants.FILE_NAME,
-								// StringConstants.FILE_PATH,
-								// StringConstants.PROJ_ID,
-								// ("package testPkg1;\n" + "\n" + "public class
-								// TestClass1 {\n"
-								// + "\n" + "}").getBytes());
-								// try {
-								// getWSManager().sendRequest(data.getRequest());
-								// } catch (ConnectException e) {
-								// // TODO Auto-generated catch block
-								// e.printStackTrace();
-								// }
-							}, null);
-					try {
-						getWSManager().sendRequest(req2, -1);
-					} catch (ConnectException e) {
-						throw new IllegalStateException("Failed to connect while attempting to log in", e);
+	
+	private void initPropertyListeners() {		
+		dataManager.getSessionStorage().addPropertyChangeListener((event) -> {
+			if (event.getPropertyName().equals(SessionStorage.USERNAME)) {
+				requestManager.fetchProjects();
+				Display.getDefault().asyncExec(() -> {
+					if (event.getOldValue() == null || !event.getOldValue().equals(event.getNewValue())) {
+						if (Window.OK == OkCancelDialog.createDialog("Do you want to auto-subscribe to subscribed projets from the last session?\n"
+								+ "This will overwrite any local changes made since the last online session.").open()) {
+							autoSubscribeForSession = true;
+						} else {
+							autoSubscribeForSession = false;
+							setAllSubscribedPrefs(false);
+						}
 					}
-				}, null);
-		getWSManager().sendRequest(req1, -1);
-
+				});
+			} else if (event.getPropertyName().equals(SessionStorage.PROJECT_LIST)) {
+				SessionStorage storage = dataManager.getSessionStorage();
+				List<Long> subscribedIdsFromPrefs = getSubscribedProjectIds();
+				Set<Long> subscribedIds = storage.getSubscribedIds();
+				if (autoSubscribeForSession) {
+					for (Long id : subscribedIdsFromPrefs) {
+						Project p = storage.getProjectById(id);
+						if (p == null) {
+							removeProjectIdFromPrefs(id);
+						} else if (!subscribedIds.contains(id)) {
+							requestManager.subscribeToProject(id);
+						}
+					}
+				}
+			}
+		});
+	}
+	
+	/**
+	 * Removes the "auto-subscribe" preference associated with the given projectID.
+	 * Should be called when either the project is no longer on the server or the
+	 * user no longer has permissions for a project.
+	 * 
+	 * @param id
+	 */
+	public void removeProjectIdFromPrefs(long id) {
+		Preferences pluginPrefs = InstanceScope.INSTANCE.getNode(Activator.PLUGIN_ID);
+		Preferences projectPrefs = pluginPrefs.node(PreferenceConstants.NODE_PROJECTS);
+		String[] projectIDs;
+		try {
+			projectIDs = projectPrefs.childrenNames();
+			for (int i = 0; i < projectIDs.length; i++) {
+				if (id == Long.parseLong(projectIDs[i])) {
```

Comment:
Is there no way to do a lookup or use a map as the underlying serialization type?